src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,220 @@
+            CancellationToken cancellationToken)
+        {
+            AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper>> lazyHandlers;
+            lock (_gate)
+            {
+                // May be called a lot.  So we use the non-allocating form of this lookup pattern.
+                lazyHandlers = cachedHandlers.GetOrAdd(
+                    messageName,
+                    static (messageName, arg) => AsyncLazy.Create(
+                        static (arg, cancellationToken) => arg.@this.ComputeHandlersAsync(arg.messageName, arg.isSolution, cancellationToken),
+                        (messageName, arg.@this, arg.isSolution)),
+                    (messageName, @this: this, isSolution));
+            }
+
+            var handlers = await lazyHandlers.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            if (handlers.Length == 0)
+                throw new InvalidOperationException($"No handler found for message {messageName}.");
+
+            if (handlers.Length > 1)
+                throw new InvalidOperationException($"Multiple handlers found for message {messageName}.");
+
+            var handler = handlers[0];
+            if (s_disabledExtensionHandlers.TryGetValue(handler, out _))
+                throw new InvalidOperationException($"Handler was disabled due to previous exception.");
+
+            try
+            {
+                var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+                var result = await ((IExtensionMessageHandlerWrapper<TArgument>)handler).ExecuteAsync(
+                    message, executeArgument, cancellationToken).ConfigureAwait(false);
+                return JsonSerializer.Serialize(result, handler.ResponseType);
+            }
+            catch (Exception ex) when (DisableHandlerAndPropagate(ex))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            bool DisableHandlerAndPropagate(Exception ex)
+            {
+                FatalError.ReportNonFatalError(ex, ErrorSeverity.Critical);
+
+                // Any exception thrown in this method is left to bubble up to the extension. But we unregister this handler
+                // from that assembly to minimize the impact.
+                s_disabledExtensionHandlers.TryAdd(handler, handler);
+                return false;
+            }
+        }
+
+        private async Task<ImmutableArray<IExtensionMessageHandlerWrapper>> ComputeHandlersAsync(
+            string messageName, bool isSolution, CancellationToken cancellationToken)
+        {
+            using var _1 = ArrayBuilder<ExtensionFolder>.GetInstance(out var extensionFolders);
+            lock (_gate)
+            {
+                foreach (var (_, extensionFolder) in _folderPathToExtensionFolder_useOnlyUnderLock)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    extensionFolders.Add(extensionFolder);
```

Comment:
I assume you tried having these be immutabledictionaries already?

Reply:
And a reason to not use `extensionFolders.Values.ToImmutable` (or w.e it was)

Reply:
So i moved away from ImmutableDict since we still needed locks (since sometimes we do multiple ops in one step atomically).  It would make some things easier.  but i felt it was about a wash.